Fixes and changes to how we handle mySQL

stealthmajr opened the pull request:
Thought PCore could use a snazzy database manager

Commit:
Fixes and changes to how we handle mySQL

Pheric commented on src/main/java/me/pheric/pcore/database/Database.java:
```diff
@@ -0,0 +1,171 @@
+     * @param username The username of this MySQL connection
+     * @param password The password of this MySQL connection
+     */
+    public Database(String jdbcUrl, String username, String password) {
```
I'd like an overload in case other options are necessary, such as specifying the port, pool size, etc.

stealthmajr replied:
Definitely! Can do. I'll most likely be allowing multiple database types, so you'll be able to use PostgreSQL, and others can use MySQL if needed.

stealthmajr replied:
Port is already being specified by the jdbcUrl. No need to add an overload for that.

Pheric replied:
Whoops, good point. It also looks like I can add my own database with `setGlobal()`

Pheric commented on src/main/java/me/pheric/pcore/database/Database.java:
```diff
@@ -0,0 +1,171 @@
+        dataSource.setJdbcUrl(jdbcUrl);
+        dataSource.setUsername(username);
+        dataSource.setPassword(password);
+        dataSource.setPoolName("Core");
```
Pool name will be "Core" even if another plugin is using the pool.

stealthmajr replied:
In my case it was a core-only scenario, I'll fix. Thanks!

Pheric commented on src/main/java/me/pheric/pcore/database/Database.java:
```diff
@@ -0,0 +1,171 @@
+            return preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            // Indicates an error
+            return -1;
```
This doesn't help much with debugging. Consider throwing an exception or printing a stack trace (maybe add a debug option that gets disabled in prod, or have the stack trace outputted to a separate logger)

stealthmajr replied:
I had a logger there but removed it because there wasn't an obtainable one in PCore.

Pheric commented on src/main/java/me/pheric/pcore/database/Database.java:
```diff
@@ -0,0 +1,171 @@
+     * @return A result object with all the data stored in it
+     */
+    Result find(String query, Object... params) {
+        // Try-resources block to automatically close the connection and prepared statement
```
I see duplicate code. Can this be broken into another method?

stealthmajr replied:
What?

Pheric replied:
I'll look into this later.

Pheric commented on src/main/java/me/pheric/pcore/database/Database.java:
```diff
@@ -0,0 +1,171 @@
+            // Check if the result set has anymore data for us
+            if (resultSet.next()) {
+                // Loop through the columns
+                for (int i = 1; i <= resultSet.getMetaData().getColumnCount(); i++) {
```
What's going on here?

stealthmajr replied:
Elaborate?

Pheric replied:
Why are you making a new `Result` instead of returning the resultset for the caller to close themself? I'll need to look into this more later, because I haven't dealt with this recently.

stealthmajr replied:
Because it's much more difficult to handle result sets than importing the results of it into another class.

Commit:
Fix some things Pheric brought up, and add lombok to the project.

Pheric reviewed (changes requested):
I'd like to see more databases supported before merging. The old system supported multiple ways of accessing Redis, MySQL/MariaDB, and PostgreSQL, and made adding new ones easier. As far as I can tell, it also was easier to configure the connection pool per-plugin, but this should be fixed with an overload.

Commit:
Minor Changes

+ Database#getSafeConnection() to allow for safe checking of the connection
+ Make more methods in the Result class
+ Implement logging for debug purposes
+ Metadata access in the Result class

Co-Authored-By: Eric Graham <pheric@users.noreply.github.com>
